25940 support credentials issuance through oid4vci

wistefan opened the pull request:
Implements the OID4VCI endpoints and the Pre-Authorized Code Grant to allow issuance of verifiable credentials 

closes #25940

wistefan committed:
Implement the oid4vci endpoints

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
Implement the oid4vci endpoints

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
Implement the oid4vci endpoints

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
Implement the oid4vci endpoints

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
implement oid4vci endpoint

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
implement oid4vci endpoint

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
implement tests

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
implement tests

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
implement tests

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
implement tests

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
add documentation and update tests

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
add documentation and update tests

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
add documentation and update tests

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
add documentation and update tests

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
code quality improvements

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
fix doc

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

babisRoutis commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -0,0 +1,402 @@
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path("credential-offer-uri")
+    public Response getCredentialOfferURI(@QueryParam("credentialId") String vcId) {
```
'credential_id' has a different meaning in VCI. Consider using 'credential_configuration_id' instead.

wistefan replied:
Fixed in [#44768c2](https://github.com/keycloak/keycloak/pull/27931/commits/44768c253324780f6aefefb4247e0a59639fb8a2)

babisRoutis commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -0,0 +1,402 @@
+            default -> throw new BadRequestException(
+                    getErrorResponse(ErrorType.UNSUPPORTED_CREDENTIAL_TYPE));
+        }
+        return Response.ok().entity(responseVO)
```
What about a fresh 'cNonce' and 'NonceExpiresIn' with the response?

This is needed to allow wallet to request again a new vc with the same access_token.

wistefan replied:
As far as  I understand, cNonce is only used in proof-binding, e.g. in the proof-object provided by the requester. Since we don't support that in that PR, it would not help in any scenario. Or do I misunderstand something here?

babisRoutis replied:
If proof is out of scope of this PR then there is no need to check proof (while processing the credential request) or to provide fresh values with the response or the error response. 

Keep in mind that for some formats that's not optional. For instance for `mso_mdoc` proof is a way to communicate the device key, that's mandatory. Also in SD-JWT-VC is highly desirable to pass a proof in order to assemble the `cnf` claim  in the issued sd-jwt to enable presentations using  key binding JWT.

wistefan replied:
I agree, but the goal really was to not let the PR explode. Proof-binding should be added in the near future, thought.

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCContextMapper.java:
```diff
@@ -0,0 +1,110 @@
+ *
+ * @author <a href="https://github.com/wistefan">Stefan Wiedemann</a>
+ */
+public class OID4VCContextMapper extends OID4VCMapper {
```
No, just because I forgot it... Thanks for mentioning:) The [jwt-profile](https://identity.foundation/jwt-vc-presentation-profile/) references the [W3C VC Datamodel](https://www.w3.org/TR/vc-data-model/), which requires @context, thus (at least to my interpretation) it is also required for jwt and sd-jwt.

wistefan replied:
Fixed in [#44768c2](https://github.com/keycloak/keycloak/pull/27931/commits/44768c253324780f6aefefb4247e0a59639fb8a2)

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCContextMapper.java:
```diff
@@ -0,0 +1,110 @@
+        return CONFIG_PROPERTIES;
+    }
+
+    public static ProtocolMapperModel create(String name) {
```
Its intended to be used in the default scopes(OID4VCLoginProtocolFactory). However, I could remove it until this one is actually used by a default scope

wistefan committed:
update pojos, make collections immutable

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
update pojos, make collections immutable

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
fix mapper initialization

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
fix mapper initialization

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/CredentialIssuer.java:
```diff
@@ -0,0 +1,105 @@
+ * @author <a href="https://github.com/wistefan">Stefan Wiedemann</a>
+ */
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class CredentialIssuer {
```
Fixed in [#44768c2](https://github.com/keycloak/keycloak/pull/27931/commits/44768c253324780f6aefefb4247e0a59639fb8a2)

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/SupportedCredential.java:
```diff
@@ -0,0 +1,215 @@
+ * {@see https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#name-credential-issuer-metadata}
+ */
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class SupportedCredential {
```
Fixed in [#44768c2](https://github.com/keycloak/keycloak/pull/27931/commits/44768c253324780f6aefefb4247e0a59639fb8a2)

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/PreAuthorizedCode.java:
```diff
@@ -0,0 +1,54 @@
+ * @author <a href="https://github.com/wistefan">Stefan Wiedemann</a>
+ */
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class PreAuthorizedCode {
```
Fixed in [#44768c2](https://github.com/keycloak/keycloak/pull/27931/commits/44768c253324780f6aefefb4247e0a59639fb8a2)

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/OID4VCClient.java:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    public OID4VCClient setSupportedVCTypes(List<SupportedCredential> supportedVCTypes) {
+        this.supportedVCTypes = supportedVCTypes;
```
Fixed in [#44768c2](https://github.com/keycloak/keycloak/pull/27931/commits/44768c253324780f6aefefb4247e0a59639fb8a2)

francis-pouatcha replied:
Any Fix here?

wistefan replied:
Sorry, now [9dd2b74](https://github.com/keycloak/keycloak/pull/27931/commits/9dd2b74b3561d8d4507b35b612872ed116b7c5e9)

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/Role.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    public Role(Set<String> names, String target) {
+        this.names = names;
```
Fixed in [#44768c2](https://github.com/keycloak/keycloak/pull/27931/commits/44768c253324780f6aefefb4247e0a59639fb8a2)

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/SupportedCredential.java:
```diff
@@ -0,0 +1,215 @@
+    }
+
+    public SupportedCredential setCryptographicBindingMethodsSupported(List<String> cryptographicBindingMethodsSupported) {
+        this.cryptographicBindingMethodsSupported = cryptographicBindingMethodsSupported;
```
Fixed in [#44768c2](https://github.com/keycloak/keycloak/pull/27931/commits/44768c253324780f6aefefb4247e0a59639fb8a2)

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/SupportedCredential.java:
```diff
@@ -0,0 +1,215 @@
+    }
+
+    public SupportedCredential setCryptographicSuitesSupported(List<String> cryptographicSuitesSupported) {
+        this.cryptographicSuitesSupported = cryptographicSuitesSupported;
```
Fixed in [#44768c2](https://github.com/keycloak/keycloak/pull/27931/commits/44768c253324780f6aefefb4247e0a59639fb8a2)

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/SupportedCredential.java:
```diff
@@ -0,0 +1,215 @@
+    }
+
+    public SupportedCredential setCredentialSigningAlgValuesSupported(List<String> credentialSigningAlgValuesSupported) {
+        this.credentialSigningAlgValuesSupported = credentialSigningAlgValuesSupported;
```
Fixed in [#44768c2](https://github.com/keycloak/keycloak/pull/27931/commits/44768c253324780f6aefefb4247e0a59639fb8a2)

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCSubjectIdMapper.java:
```diff
@@ -0,0 +1,105 @@
+
+    @Override
+    public void setClaimsForSubject(Map<String, Object> claims, UserSessionModel userSessionModel) {
+        claims.put("id", mapperModel.getConfig().getOrDefault(ID_KEY, String.format("urn:uuid:%s", UUID.randomUUID())));
```
No. Id is the field in the credential, ID_KEY is the key of the configuration property.

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCTypeMapper.java:
```diff
@@ -0,0 +1,109 @@
+
+    public OID4VCTypeMapper() {
+        super();
+        ProviderConfigProperty vcTypePropertyNameConfig = new ProviderConfigProperty();
```
TYPE_KEY is the value for the key of the property, not the value itself. We could use "VerifiableCredential" as default value, thought.

wistefan replied:
fixed in [#087a816](https://github.com/keycloak/keycloak/pull/27931/commits/087a8162c44b877a1a097085b17ad006a8d7c6ed)

wistefan committed:
more review cleanups

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
update doc reference

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCStaticClaimMapper.java:
```diff
@@ -0,0 +1,112 @@
+        CONFIG_PROPERTIES.add(claimValueConfig);
+    }
+
+    public static ProtocolMapperModel create(String mapperName, String propertyName, String value) {
```
fixed in [#087a816](https://github.com/keycloak/keycloak/pull/27931/commits/087a8162c44b877a1a097085b17ad006a8d7c6ed)

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCTypeMapper.java:
```diff
@@ -0,0 +1,109 @@
+        if (verifiableCredential.getType() != null) {
+            types = new HashSet<>(verifiableCredential.getType());
+        }
+        types.add(mapperModel.getConfig().get(TYPE_KEY));
```
fixed in [#087a816](https://github.com/keycloak/keycloak/pull/27931/commits/087a8162c44b877a1a097085b17ad006a8d7c6ed)

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCTypeMapper.java:
```diff
@@ -0,0 +1,109 @@
+        return CONFIG_PROPERTIES;
+    }
+
+    public static ProtocolMapperModel create(String name) {
```
fixed in [#087a816](https://github.com/keycloak/keycloak/pull/27931/commits/087a8162c44b877a1a097085b17ad006a8d7c6ed)

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/CredentialsOffer.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    public CredentialsOffer setCredentialConfigurationIds(List<String> credentialConfigurationIds) {
+        this.credentialConfigurationIds = credentialConfigurationIds;
```
fixed in [#087a816](https://github.com/keycloak/keycloak/pull/27931/commits/087a8162c44b877a1a097085b17ad006a8d7c6ed)

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/CredentialIssuer.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    public CredentialIssuer setCredentialsSupported(Map<String, SupportedCredential> credentialsSupported) {
+        this.credentialsSupported = credentialsSupported;
```
fixed in [#087a816](https://github.com/keycloak/keycloak/pull/27931/commits/087a8162c44b877a1a097085b17ad006a8d7c6ed)

wistefan committed:
make it immutable

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
fix merge conflicts

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
fix merge conflicts

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
fix test

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

jonkoops commented:
Can we resolve the comments on this PR that have been fixed? It's kind of hard to get an overview at the moment.

wistefan commented:
I set the once from @francis-pouatcha to resolved, since he already approved. @babisRoutis could you check your comments, please?

babisRoutis commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java:
```diff
@@ -0,0 +1,118 @@
+                .setCredentialIssuer(getIssuer(keycloakSession.getContext()))
+                .setCredentialEndpoint(getCredentialsEndpoint(keycloakSession.getContext()))
+                .setCredentialsSupported(getSupportedCredentials(keycloakSession))
+                .setAuthorizationServers(List.of(getIssuer(keycloakSession.getContext())));
```
The VCI specifications states:

> If this parameter is omitted, the entity providing the Credential Issuer is also acting as the Authorization Server, i.e., the Credential Issuer's identifier is used to obtain the Authorization Server metadata. The actual OAuth 2.0 Authorization Server metadata is obtained from the oauth-authorization-server well-known location as defined in Section 3 of [[RFC8414](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#RFC8414)].

To my understanding this means that  KC could skip setting the authorization servers list. Even if it is set, I see no harm though.

wistefan replied:
I agree in general. According to the spec, it would be optional and can be empty in our case. However, there are a number of wallets out there, that handle it as mandatory and fail without. Especially the once listed under [the EBSI-conformance list](https://ec.europa.eu/digital-building-blocks/sites/display/EBSI/Conformant+wallets) are susceptible to that, since the conformance test provides the authorization_servers. Thus, I decided to set it, since that is still valid with the spec.

babisRoutis replied:
@wistefan I think draft 13 will "break" many wallet implementations :smiley: 
This field used do be a single value (not a list).  

Anyway. No problem if set.

babisRoutis commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/VerifiableCredential.java:
Should this class `VerifiableCredential` & `CredentialSubject` be part of the VCI model?
Haven't checked its usage but I have the impression that they both are related to W3C VC DM

wistefan replied:
I think so. At least the jwt-profiles also reference the W3C VC DM. And we are using them for communicating between the IssuerEndpoint class and the concrete implementations of the SigningServices

babisRoutis commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/ProofType.java:
```diff
@@ -0,0 +1,40 @@
+ */
+public enum ProofType {
+
+    JWT("jwt"),
```
I think there is also the `cwt` case

babisRoutis commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/ProofType.java:
```diff
@@ -0,0 +1,40 @@
+public enum ProofType {
+
+    JWT("jwt"),
+    LD_PROOF("ld_proof");
```
Is this `ld_proof` correct? I think it is `ldp_vp`

wistefan replied:
fixed in [#efc83d5](https://github.com/keycloak/keycloak/pull/27931/commits/efc83d575b68430da840ad8cc53ce2eb0ffb887e)

babisRoutis commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/Proof.java:
```diff
@@ -0,0 +1,54 @@
+    @JsonProperty("proof_type")
+    private ProofType proofType;
+
+    private String jwt;
```
Is `Proof` only for jwt? Should it be renamed to `JwtProof`?

Otherwise it cannot represent `cwt` and ld_vp proof

wistefan replied:
fixed in [#efc83d5](https://github.com/keycloak/keycloak/pull/27931/commits/efc83d575b68430da840ad8cc53ce2eb0ffb887e)

wistefan committed:
fix proof related values

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
fix dependency issues

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

ahus1 commented:
@wistefan - I see you are rebasing your PR regularly. Please only rebase your PR when there are conflicts, or a maintainer asks you to rebase your PR. This avoids running unnecessary CI jobs which might delay others, and also avoid sending out notifications to people watching this PR, and also avoid invalidating previous reviews. Thanks!

tnorimat commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/TxCode.java:
```diff
@@ -0,0 +1,50 @@
+package org.keycloak.protocol.oid4vc.model;
```
Could you add header comments the same as other newly created files?

wistefan replied:
Fixed in [7beca5e](https://github.com/keycloak/keycloak/pull/27931/commits/7beca5e54dc584c6726a73ca82109f0fc3d22124)

tnorimat replied:
Thank you.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/grants/PreAuthorizedCodeGrantType.java:
```diff
@@ -0,0 +1,117 @@
+package org.keycloak.protocol.oidc.grants;
```
Could you add header comments the same as other newly created files?

wistefan replied:
[7beca5e](https://github.com/keycloak/keycloak/pull/27931/commits/7beca5e54dc584c6726a73ca82109f0fc3d22124)

tnorimat replied:
Thank you.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/grants/PreAuthorizedCodeGrantTypeFactory.java:
```diff
@@ -0,0 +1,38 @@
+package org.keycloak.protocol.oidc.grants;
```
Could you add header comments the same as other newly created files?

wistefan replied:
[7beca5e](https://github.com/keycloak/keycloak/pull/27931/commits/7beca5e54dc584c6726a73ca82109f0fc3d22124)

tnorimat replied:
Thank you.

tnorimat commented on services/src/test/java/org/keycloak/protocol/oid4vc/OID4VCClientRegistrationProviderTest.java:
```diff
@@ -0,0 +1,115 @@
+package org.keycloak.protocol.oid4vc;
```
Could you add header comments the same as other newly created files?

wistefan replied:
Fixed in  [7beca5e](https://github.com/keycloak/keycloak/pull/27931/commits/7beca5e54dc584c6726a73ca82109f0fc3d22124)

tnorimat replied:
Thank you.

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/PreAuthorizedGrantTest.java:
```diff
@@ -0,0 +1,121 @@
+package org.keycloak.testsuite.oauth;
```
Could you add header comments the same as other newly created files?

tnorimat replied:
Thank you.

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCIssuerEndpointTest.java:
```diff
@@ -0,0 +1,506 @@
+package org.keycloak.testsuite.oid4vc.issuance.signing;
```
Could you add header comments the same as other newly created files?

wistefan replied:
Fixed in  [7beca5e](https://github.com/keycloak/keycloak/pull/27931/commits/7beca5e54dc584c6726a73ca82109f0fc3d22124)

tnorimat replied:
Thank you.

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCIssuerWellKnownProviderTest.java:
```diff
@@ -0,0 +1,70 @@
+package org.keycloak.testsuite.oid4vc.issuance.signing;
```
Could you add header comments the same as other newly created files?

wistefan replied:
Fixed in  [7beca5e](https://github.com/keycloak/keycloak/pull/27931/commits/7beca5e54dc584c6726a73ca82109f0fc3d22124)

tnorimat replied:
Thank you.

tnorimat reviewed (changes requested):
@wistefan Thank you for your work.
Some CI tests failed. Before reviewing the PR, could you revise codes to pass these tests?

bucchi commented:
Hi,

I have run OID4VCIssuerEndpointTest on “25940-support-credentials-issuance-through-oid4vci” branch.
I found out something wrong, so let me share that here.

OID4VCIssuerEndpointTest , as it is committed, goes well actually.
But I have tried to dig deeper and added some assert statement(*) in requestOffer method for checking  inside vc. 
Then it looked something going wrong.

Some claims, which is supposed to be  in vc object, is missing like below.

> {
>   "nbf" : 1712172643,
>   "jti" : "urn:uuid:b9ca7966-5841-4087-b9ee-526e19bad68b",
>   "iss" : "did:web:test.org",
>   "vc" : {
>     "type" : [ "VerifiableCredential" ],
>     "issuer" : "did:web:test.org",
>     "issuanceDate" : 1712172643192,
>     "credentialSubject" : { }
>   }

Maybe  it is because mapper is not called properly, where it should be.

OID4VCIssuerEndpoint with getCredential method   
try to fetch OID4VCMapper from session with 
“session.getAllProviders(OID4VCMapper.class)”.
But actually it gets empty object and as a result no mapper gets called there.


(*) Added assert statements are something like below at the very last of the requestOffer method in OID4VCIssuerEndpointTest.
```
        JsonWebToken jsonWebToken = TokenVerifier.create((String) credentialResponse.getCredential(), JsonWebToken.class).getToken();
        assertEquals("did:web:test.org", jsonWebToken.getIssuer());

        VerifiableCredential credential = new ObjectMapper().convertValue(jsonWebToken.getOtherClaims().get("vc"), VerifiableCredential.class);
        assertEquals(List.of("VerifiableCredential"), credential.getType());
        assertEquals(URI.create("did:web:test.org"), credential.getIssuer());
        assertEquals("john@email.cz", credential.getCredentialSubject().getClaims().get("email"));

```

wistefan commented:
> @wistefan - I see you are rebasing your PR regularly. Please only rebase your PR when there are conflicts, or a maintainer asks you to rebase your PR. This avoids running unnecessary CI jobs which might delay others, and also avoid sending out notifications to people watching this PR, and also avoid invalidating previous reviews. Thanks!

Sorry for the noise, I've seen some tests failing only in the CI after rebasing and then was "hoping" they resolve by updating. I'm now able to reproduce them locally, too and work there.

wistefan committed:
fix mapper instantiation, add comments

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan committed:
restart test after feature activation

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan commented:
> > Hi,
> > I have run OID4VCIssuerEndpointTest on “25940-support-credentials-issuance-through-oid4vci” branch. I found out something wrong, so let me share that here.
> > OID4VCIssuerEndpointTest , as it is committed, goes well actually. But I have tried to dig deeper and added some assert statement(*) in requestOffer method for checking inside vc. Then it looked something going wrong.
> > Here is what happens: Some claims, which is supposed to be in vc object, is missing like below.
> > > {
> > > "nbf" : 1712172643,
> > > "jti" : "urn:uuid:b9ca7966-5841-4087-b9ee-526e19bad68b",
> > > "iss" : "did:web:test.org",
> > > "vc" : {
> > > "type" : [ "VerifiableCredential" ],
> > > "issuer" : "did:web:test.org",
> > > "issuanceDate" : 1712172643192,
> > > "credentialSubject" : { }
> > > }
> > 
> > 
> > Maybe it is because mapper is not called properly, where it should be.
> > OID4VCIssuerEndpoint with getCredential method tries to fetch OID4VCMapper from session with “session.getAllProviders(OID4VCMapper.class)”. But actually it gets empty object and as a result no mapper gets called there.
> > (*) FYI, added assert statements are something like below at the very last of the requestOffer method in OID4VCIssuerEndpointTest.
> > ```
> >         JsonWebToken jsonWebToken = TokenVerifier.create((String) credentialResponse.getCredential(), JsonWebToken.class).getToken();
> >         assertEquals("did:web:test.org", jsonWebToken.getIssuer());
> > 
> >         VerifiableCredential credential = new ObjectMapper().convertValue(jsonWebToken.getOtherClaims().get("vc"), VerifiableCredential.class);
> >         assertEquals(List.of("VerifiableCredential"), credential.getType());
> >         assertEquals(URI.create("did:web:test.org"), credential.getIssuer());
> >         assertEquals("john@email.cz", credential.getCredentialSubject().getClaims().get("email"));
> > ```
> 
> Hi, thank you, I will check whats happening.

@bucchi 
I fixed it in  [7beca5e](https://github.com/keycloak/keycloak/pull/27931/commits/7beca5e54dc584c6726a73ca82109f0fc3d22124) and added your assertions.

ahus1 commented:
@wistefan - thank, now I understand you situation better. If you run into problems running the test locally, I'm happy to lend a hand for that. Sorry for my comment here which was more harsh than it should have been.

To use the GitHub actions to run the tests, you can start a PR as "draft" and then only those people you mention will get notifications. The notifications to all code owners are then sent out only once you mark it "ready for review". 

If you start it as "read to review", and then later switch it to "draft", maintainers will still get notifications, but they can unsubscribe until you switch it to "read for review again". 

Best,
Alexander

wistefan commented:
> @wistefan Thank you for your work. Some CI tests failed. Before reviewing the PR, could you revise codes to pass these tests?

@tnorimat I fixed the tests in [3372d3c](https://github.com/keycloak/keycloak/pull/27931/commits/3372d3c05b74ca78543139c87b84c87881d250fe). The issue happend, because enabling the OID4VCI feature for the test worked on the default test environment(with undertow) but on quarkus(the auth server used in the CI) requires a restart.

tnorimat commented:
@wistefan Hello, I am back from my business trip in Europe. I will review the PR in this week.

mposolda commented on testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/rest/TestingResourceProvider.java:
```diff
@@ -224,7 +226,7 @@ public Response revertTestingInfinispanTimeService() {
     @Path("/get-client-sessions-count")
     @Produces(MediaType.APPLICATION_JSON)
-    public Integer getClientSessionsCountInUserSession(@QueryParam("realm") final String name, @QueryParam("session") final String sessionId) {
+    public Integer getClientSessionsCountInssion(@QueryParam("realm") final String name, @QueryParam("session") final String sessionId) {
```
Minor: Not sure why this method was updated? New name does not makes much sense :-) (even if it probably does not break anything...). I guess it was a typo?

wistefan replied:
Not sure how that happend, was a mistake. Fixed in [a65798d](https://github.com/keycloak/keycloak/pull/27931/commits/a65798dfaa126bc4edbcc254a77affa5c24816bc)

mposolda commented on services/src/main/java/org/keycloak/protocol/oid4vc/OID4VCLoginProtocolFactory.java:
```diff
@@ -0,0 +1,173 @@
+        ClientScopeModel naturalPersonScope = KeycloakModelUtils.getClientScopeByName(newRealm, "natural_person");
+        if (naturalPersonScope == null) {
+            LOGGER.debug("Add natural person scope");
+            naturalPersonScope = newRealm.addClientScope("natural_person");
```
I am thinking if it makes sense to name this client scope as `oid4vc_natural_person` to make it more clear that it is about `oid4vc` ? But not strong opinion on this as protocol is visible anyway in the table of client scopes. @tnorimat WDYT?

mposolda commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCMapper.java:
```diff
@@ -0,0 +1,122 @@
+ *
+ * @author <a href="https://github.com/wistefan">Stefan Wiedemann</a>
+ */
+public abstract class OID4VCMapper implements ProtocolMapper {
```
Should this one also implement `OID4VCEnvironmentProviderFactory` to make sure that protocol mappers are available just for `oid4vc` protocol? I assume that it is not strictly needed, but at the same time, it probably won't hurt?

wistefan replied:
Yes, I added it in [a65798d](https://github.com/keycloak/keycloak/pull/27931/commits/a65798dfaa126bc4edbcc254a77affa5c24816bc)

mposolda reviewed (commented):
@wistefan This is great stuff!

The PR is a bit bigger and I did not reviewed extensively all the aspects of the protocol. Relying on the review of @tnorimat, @francis-pouatcha and @babisRoutis for that. Thanks to all of you for the review!

But overally, seems to me as great stuff. I've added some comments inline (none of them is probably blocker of this PR and some are more a questions, but if you can address them, it can be nice).

Besides that, I hope to merge once @tnorimat reviews and approves this PR.

wistefan committed:
minor improvements

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/PreAuthorizedGrantTest.java:
```diff
@@ -0,0 +1,138 @@
+import static org.junit.Assert.assertEquals;
+
+@EnableFeature(value = Profile.Feature.OID4VC_VCI, skipRestart = true)
+public class PreAuthorizedGrantTest extends AbstractTestRealmKeycloakTest {
```
It seems that tests do not access Credential Endpoint to get a verifiable credential  in return for an access token. I think we need to test this point.

Would you intend not to do such the tests in the PR, but do suche tests in the follwoing PR after the PR is mereged?

wistefan replied:
I see the actual credential retrieval rather as part of the OID4VCIssuerEndpointTest(where I do test exactly what you mentioned).

tnorimat reviewed (commented):
@wistefan Hello,
I partially reviewed the PR and add some comments Could you check them?

Also, I need more time to complete the PR because the size of codes of the PR is big as @mposolda  said. I will do my best for it in the week.

wistefan committed:
fix header doc, use ClientSessionContext, code cleanups

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan commented on services/src/main/java/org/keycloak/protocol/oidc/grants/PreAuthorizedCodeGrantType.java:
```diff
@@ -0,0 +1,134 @@
+                    Response.Status.BAD_REQUEST);
+        }
+        AuthenticatedClientSessionModel clientSession = result.getClientSession();
+        DefaultClientSessionContext sessionContext = DefaultClientSessionContext.fromClientSessionAndScopeParameter(clientSession,
```
Fixed in [bbdb87f](https://github.com/keycloak/keycloak/pull/27931/commits/bbdb87f6cbdeb6fc55aeebf7b826a353570e7ab9)

tnorimat replied:
Thank you.

wistefan commented on services/src/main/java/org/keycloak/protocol/oidc/grants/PreAuthorizedCodeGrantType.java:
```diff
@@ -0,0 +1,134 @@
+        session.getContext().setClient(result.getClientSession().getClient());
+
+
+        // set the client as retrieved from the pre-authorized session
```
Fixed in [bbdb87f](https://github.com/keycloak/keycloak/pull/27931/commits/bbdb87f6cbdeb6fc55aeebf7b826a353570e7ab9)

wistefan commented on services/src/main/java/org/keycloak/protocol/oidc/grants/PreAuthorizedCodeGrantType.java:
```diff
@@ -0,0 +1,134 @@
+import org.keycloak.events.EventType;
+import org.keycloak.models.AuthenticatedClientSessionModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.protocol.oid4vc.OID4VCClientRegistrationProvider;
```
Fixed in [bbdb87f](https://github.com/keycloak/keycloak/pull/27931/commits/bbdb87f6cbdeb6fc55aeebf7b826a353570e7ab9)

tnorimat replied:
Thank you.

wistefan commented on services/src/main/java/org/keycloak/protocol/oidc/grants/PreAuthorizedCodeGrantType.java:
```diff
@@ -0,0 +1,134 @@
+     *
+     * @param session                    - keycloak session to be used
+     * @param authenticatedClientSession - client session to be persisted
+     * @param expirationTime             - expiration time of the code, the code should be short-lifed
```
Fixed in [bbdb87f](https://github.com/keycloak/keycloak/pull/27931/commits/bbdb87f6cbdeb6fc55aeebf7b826a353570e7ab9)

tnorimat replied:
Thank you.

wistefan commented on services/src/main/resources/META-INF/services/org.keycloak.protocol.oid4vc.issuance.signing.VCSigningServiceProviderFactory:
```diff
@@ -0,0 +1,19 @@
+#
+# Copyright 2016 Red Hat, Inc. and/or its affiliates
```
Fixed in [bbdb87f](https://github.com/keycloak/keycloak/pull/27931/commits/bbdb87f6cbdeb6fc55aeebf7b826a353570e7ab9)

tnorimat replied:
Thank you.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -0,0 +1,388 @@
+                .orElseThrow(
+                        () -> {
+                            LOGGER.debugf("Unsupported credential %s was requested.", requestedCredential);
+                            throw new BadRequestException(getErrorResponse(ErrorType.UNSUPPORTED_CREDENTIAL_TYPE));
```
`throw new BadRequestException(getErrorResponse(ErrorType.UNSUPPORTED_CREDENTIAL_TYPE));`

might be

`return new BadRequestException(getErrorResponse(ErrorType.UNSUPPORTED_CREDENTIAL_TYPE));`

wistefan replied:
Fixed in [0d88edc](https://github.com/keycloak/keycloak/pull/27931/commits/0d88edc5952e78aed03bc8acaaaa3f4674de7b5a)

tnorimat replied:
Thank you.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -0,0 +1,388 @@
+import org.keycloak.protocol.oid4vc.model.SupportedCredentialConfiguration;
+import org.keycloak.protocol.oid4vc.model.VerifiableCredential;
+import org.keycloak.protocol.oidc.grants.PreAuthorizedCodeGrantType;
+import org.keycloak.provider.ProviderFactory;
```
It seems that it is not used.

wistefan replied:
Fixed in [0d88edc](https://github.com/keycloak/keycloak/pull/27931/commits/0d88edc5952e78aed03bc8acaaaa3f4674de7b5a)

tnorimat replied:
Thank you.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -0,0 +1,388 @@
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
```
It seems that it is not used.

wistefan replied:
Fixed in [0d88edc](https://github.com/keycloak/keycloak/pull/27931/commits/0d88edc5952e78aed03bc8acaaaa3f4674de7b5a)

tnorimat commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -0,0 +1,388 @@
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path("credential-offer-uri")
+    public Response getCredentialOfferURI(@QueryParam("credential_configuration_id") String vcId) {
```
The specification of oid4vci dose not specify credential offer endpoint's interface.
Therefore, is it keycloak's specific interface ?

wistefan replied:
Yes. In order to create a qr, we need a way to get an offer-uri. Thus I added that endpoint.

tnorimat replied:
Thank you for the clarification.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -0,0 +1,388 @@
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path(CREDENTIAL_OFFER_PATH + "{nonce}")
+    public Response getCredentialOffer(@PathParam("nonce") String nonce) {
```
The specification of oid4vci dose not specify credential offer endpoint's interface.
Therefore, is it keycloak's specific interface ?

wistefan replied:
There is a non-normative example, that I used for that https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#section-4.1.3 
So, yes it probably has to be seen as keycloak specific

tnorimat replied:
Thank you for the clarification.

tnorimat commented on testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/util/OAuthClient.java:
```diff
@@ -58,6 +58,8 @@
 import org.keycloak.models.utils.KeycloakModelUtils;
 import org.keycloak.protocol.oidc.OIDCLoginProtocol;
 import org.keycloak.protocol.oidc.OIDCLoginProtocolService;
+import org.keycloak.protocol.oidc.grants.PreAuthorizedCodeGrantType;
```
It seems that it is not used.

wistefan replied:
Fixed in [42f5903](https://github.com/keycloak/keycloak/pull/27931/commits/42f5903f054b931386c5c249d4f2990c99ed4673)

tnorimat commented on testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/util/OAuthClient.java:
```diff
@@ -790,6 +793,27 @@ public AccessTokenResponse doClientCredentialsGrantAccessTokenRequest(String cli
         }
     }
 
+    public AccessTokenResponse doPreauthorizedTokenRequest(String preAuthorizedCode) throws Exception {
```
Why it is not used in `PreAuthorizedGrantTest` ?

wistefan replied:
Because I wanted to make the usage of the api and its parameters more explicit in the test. E.g. for testing the PreAuthorizedToken Feature, I build the requests in the test, for using the feature in other tests(e.g. the OID4VCI tests), I use the helper method here.

tnorimat replied:
Thank you for the clarification.

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/JwtSigningServiceTest.java:
```diff
@@ -53,7 +52,7 @@
 
 
-public class JwtSigningServiceTest extends SigningServiceTest {
+public class JwtSigningServiceTest extends OID4VCTest {
```
L76 and L78 are opposite.

wistefan replied:
Amazing, how are you able to spot such thing!?:) Fixed in [42f5903](https://github.com/keycloak/keycloak/pull/27931/commits/42f5903f054b931386c5c249d4f2990c99ed4673)

tnorimat replied:
Thank you.

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCTest.java:
```diff
@@ -19,18 +19,30 @@
 import org.keycloak.crypto.KeyWrapper;
 import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.utils.KeycloakModelUtils;
+import org.keycloak.protocol.oid4vc.OID4VCClientRegistrationProviderFactory;
```
It seems that it is not used.

wistefan replied:
Fixed in [42f5903](https://github.com/keycloak/keycloak/pull/27931/commits/42f5903f054b931386c5c249d4f2990c99ed4673)

tnorimat replied:
Thank you.

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCTest.java:
```diff
@@ -19,18 +19,30 @@
+import org.keycloak.representations.idm.UserRepresentation;
 import org.keycloak.testsuite.AbstractTestRealmKeycloakTest;
+import org.keycloak.testsuite.arquillian.annotation.EnableFeature;
+import org.keycloak.testsuite.util.RoleBuilder;
```
It seems that it is not used.

wistefan replied:
Fixed in [42f5903](https://github.com/keycloak/keycloak/pull/27931/commits/42f5903f054b931386c5c249d4f2990c99ed4673)

tnorimat replied:
Thank you.

tnorimat reviewed (commented):
@wistefan I completed the review of the PR. I added some comments. Could you check them?

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/OID4VCClientRegistrationProvider.java:
```diff
@@ -0,0 +1,169 @@
+ *
+ * @author <a href="https://github.com/wistefan">Stefan Wiedemann</a>
+ */
+public class OID4VCClientRegistrationProvider extends AbstractClientRegistrationProvider {
```
Hi, I´m not sure if I fully understand. Beside the supportedVCTypes, it also adds the did of the client. Its used to provide the ability to create credentials focused for clients supporting OID4VC standards, not neccessarily OIDC. Mixing both might be confusing, especially if one only supports one of them.

tnorimat replied:
Thank you for the clarification.

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/OID4VCLoginProtocolFactory.java:
```diff
@@ -0,0 +1,173 @@
+
+    @Override
+    public LoginProtocol create(KeycloakSession session) {
+        return null;
```
We currently support to login  with oid4vp in keycloak, thus no such protocl is implemented. As far as I understand this means returning null here is ok.

tnorimat replied:
Thank you for the clarification.

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/CredentialOfferURI.java:
```diff
@@ -0,0 +1,49 @@
+ * @author <a href="https://github.com/wistefan">Stefan Wiedemann</a>
+ */
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class CredentialOfferURI {
```
I decided to build it that way, since the retrieval endpoint for that credential_offer_uri is not defined in the spec, just the way how it should be shown to a  wallet(from my understanding). We use it to generate the string in the frontend for the qr from it. However, we could add a plain/text endpoint, too, if you think its valuable

tnorimat replied:
Thank you for the clarification.

wistefan committed:
update error types

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/ErrorType.java:
```diff
@@ -0,0 +1,44 @@
+ *
+ * @author <a href="https://github.com/wistefan">Stefan Wiedemann</a>
+ */
+public enum ErrorType {
```
Probably something changed to an older draft again, I'll updated [940c4af](https://github.com/keycloak/keycloak/pull/27931/commits/940c4afd83cd32c449c67463a5e54b6155303ba2)

tnorimat replied:
Thank you.

wistefan committed:
cleanups

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/OID4VCClientRegistrationProviderFactory.java:
```diff
@@ -0,0 +1,68 @@
+
+/**
+ * Implementation of the {@link ClientRegistrationProviderFactory} to integrate the OID4VC protocols with
+ * Keycloaks client-registration.
```
Fixed in [940c4af](https://github.com/keycloak/keycloak/pull/27931/commits/940c4afd83cd32c449c67463a5e54b6155303ba2)

tnorimat replied:
Thank you.

wistefan commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java:
```diff
@@ -130,7 +133,9 @@ public Response processGrantRequest() {
 
-        if (!grantType.equals(OAuth2Constants.UMA_GRANT_TYPE)) {
+        if (!grantType.equals(OAuth2Constants.UMA_GRANT_TYPE)
+                // pre-authorized grants are not necessarily used by known clients.
```
Its not mentioned anywhere(as far as I know). However, its my logical conclusion. When the pre-authorized grant comes through a qr-scanned offer to a wallet, the wallet is the client but cannot be known by the issuer in advance.

tnorimat replied:
Thank you for the clarification.

wistefan committed:
more cleanups

Signed-off-by: Stefan Wiedemann <wistefan@googlemail.com>

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/PreAuthorizedGrant.java:
```diff
@@ -0,0 +1,44 @@
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class PreAuthorizedGrant {
+
+    @JsonProperty("urn:ietf:params:oauth:grant-type:pre-authorized_code")
```
Fixed in [0d88edc](https://github.com/keycloak/keycloak/pull/27931/commits/0d88edc5952e78aed03bc8acaaaa3f4674de7b5a)

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/SupportedCredentialConfiguration.java:
```diff
@@ -0,0 +1,216 @@
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class SupportedCredentialConfiguration {
+
+    private static final String DOT_SEPERATOR = ".";
```
Fixed in [0d88edc](https://github.com/keycloak/keycloak/pull/27931/commits/0d88edc5952e78aed03bc8acaaaa3f4674de7b5a)

tnorimat replied:
Thank you.

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/model/SupportedCredentialConfiguration.java:
```diff
@@ -0,0 +1,216 @@
+    @JsonIgnore
+    private static final String CRYPTOGRAPHIC_BINDING_METHODS_SUPPORTED_KEY = "cryptographic_binding_methods_supported";
+    @JsonIgnore
+    private static final String CRYPTOGRAPHIC_SUITES_SUPPORTED_KEY = "cryptographic_suites_supported";
```
Fixed in [0d88edc](https://github.com/keycloak/keycloak/pull/27931/commits/0d88edc5952e78aed03bc8acaaaa3f4674de7b5a)

tnorimat replied:
Thank you.

wistefan commented on services/src/main/java/org/keycloak/protocol/oidc/OIDCWellKnownProvider.java:
```diff
@@ -33,6 +33,8 @@
 import org.keycloak.models.RealmModel;
 import org.keycloak.protocol.oidc.endpoints.AuthorizationEndpoint;
 import org.keycloak.protocol.oidc.endpoints.TokenEndpoint;
+import org.keycloak.protocol.oidc.grants.PreAuthorizedCodeGrantType;
```
Fixed in [0d88edc](https://github.com/keycloak/keycloak/pull/27931/commits/0d88edc5952e78aed03bc8acaaaa3f4674de7b5a)

tnorimat replied:
Thank you.

wistefan commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java:
```diff
@@ -33,6 +33,8 @@
 import org.keycloak.protocol.oidc.OIDCLoginProtocol;
 import org.keycloak.protocol.oidc.TokenManager;
 import org.keycloak.protocol.oidc.grants.OAuth2GrantType;
+import org.keycloak.protocol.oidc.grants.PreAuthorizedCodeGrantType;
```
Fixed in [0d88edc](https://github.com/keycloak/keycloak/pull/27931/commits/0d88edc5952e78aed03bc8acaaaa3f4674de7b5a)

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCContextMapper.java:
```diff
@@ -0,0 +1,98 @@
+package org.keycloak.protocol.oid4vc.issuance.mappers;
+
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.ProtocolMapperModel;
```
Fixed in [0d88edc](https://github.com/keycloak/keycloak/pull/27931/commits/0d88edc5952e78aed03bc8acaaaa3f4674de7b5a)

tnorimat replied:
Thank you.

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCContextMapper.java:
```diff
@@ -0,0 +1,98 @@
+import org.keycloak.models.ProtocolMapperModel;
+import org.keycloak.models.UserSessionModel;
+import org.keycloak.protocol.ProtocolMapper;
+import org.keycloak.protocol.oid4vc.OID4VCLoginProtocolFactory;
```
Fixed in [0d88edc](https://github.com/keycloak/keycloak/pull/27931/commits/0d88edc5952e78aed03bc8acaaaa3f4674de7b5a)

tnorimat replied:
Thank you.

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCMapper.java:
```diff
@@ -0,0 +1,123 @@
+        supportedCredentialsConfig.setType(ProviderConfigProperty.STRING_TYPE);
+        supportedCredentialsConfig.setLabel("Supported Credential Types");
+        supportedCredentialsConfig.setDefaultValue("VerifiableCredential");
+        supportedCredentialsConfig.setHelpText(
```
Fixed in [0d88edc](https://github.com/keycloak/keycloak/pull/27931/commits/0d88edc5952e78aed03bc8acaaaa3f4674de7b5a)

tnorimat replied:
Thank you.

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCStaticClaimMapper.java:
```diff
@@ -0,0 +1,99 @@
+package org.keycloak.protocol.oid4vc.issuance.mappers;
+
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.ProtocolMapperModel;
```
Fixed in [0d88edc](https://github.com/keycloak/keycloak/pull/27931/commits/0d88edc5952e78aed03bc8acaaaa3f4674de7b5a)

tnorimat replied:
Thank you.

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCStaticClaimMapper.java:
```diff
@@ -0,0 +1,99 @@
+import org.keycloak.models.ProtocolMapperModel;
+import org.keycloak.models.UserSessionModel;
+import org.keycloak.protocol.ProtocolMapper;
+import org.keycloak.protocol.oid4vc.OID4VCLoginProtocolFactory;
```
Fixed in [0d88edc](https://github.com/keycloak/keycloak/pull/27931/commits/0d88edc5952e78aed03bc8acaaaa3f4674de7b5a)

tnorimat replied:
Thank you.

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCStaticClaimMapper.java:
```diff
@@ -0,0 +1,99 @@
+import org.keycloak.provider.ProviderConfigProperty;
+
+import java.util.ArrayList;
+import java.util.HashMap;
```
Fixed in [0d88edc](https://github.com/keycloak/keycloak/pull/27931/commits/0d88edc5952e78aed03bc8acaaaa3f4674de7b5a)

tnorimat replied:
Thank you.

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCSubjectIdMapper.java:
```diff
@@ -0,0 +1,103 @@
+import org.keycloak.models.ProtocolMapperModel;
+import org.keycloak.models.UserSessionModel;
+import org.keycloak.protocol.ProtocolMapper;
+import org.keycloak.protocol.oid4vc.OID4VCClientRegistrationProviderFactory;
```
Fixed in [0d88edc](https://github.com/keycloak/keycloak/pull/27931/commits/0d88edc5952e78aed03bc8acaaaa3f4674de7b5a)

tnorimat replied:
Thank you.

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCTypeMapper.java:
```diff
@@ -0,0 +1,98 @@
+package org.keycloak.protocol.oid4vc.issuance.mappers;
+
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.ProtocolMapperModel;
```
Fixed in [0d88edc](https://github.com/keycloak/keycloak/pull/27931/commits/0d88edc5952e78aed03bc8acaaaa3f4674de7b5a)

tnorimat replied:
Thank you.

wistefan commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCTypeMapper.java:
```diff
@@ -0,0 +1,98 @@
+import org.keycloak.provider.ProviderConfigProperty;
+
+import java.util.ArrayList;
+import java.util.HashMap;
```
Fixed in [0d88edc](https://github.com/keycloak/keycloak/pull/27931/commits/0d88edc5952e78aed03bc8acaaaa3f4674de7b5a)

tnorimat replied:
Thank you.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCContextMapper.java:
```diff
@@ -0,0 +1,98 @@
+import org.keycloak.provider.ProviderConfigProperty;
+
+import java.util.ArrayList;
+import java.util.HashMap;
```
Thank you.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCTypeMapper.java:
```diff
@@ -0,0 +1,98 @@
+import org.keycloak.models.ProtocolMapperModel;
+import org.keycloak.models.UserSessionModel;
+import org.keycloak.protocol.ProtocolMapper;
+import org.keycloak.protocol.oid4vc.OID4VCLoginProtocolFactory;
```
Thank you.

tnorimat reviewed (approved):
LGTM.

tnorimat commented:
@wistefan Thank you very much for your work.
@mposolda Could you re-check the PR and merge it if it is OK?

mposolda reviewed (approved):
@wistefan @tnorimat @francis-pouatcha @babisRoutis Thanks everyone for the update and review of this PR!
